### ISSUE TITLE
Question about KL term

### ISSUE THREAD

**Issue by SimengSun** [tatsu-lab/alpaca_farm#67]
Hi,

Thank you for the great work! I am a novice in RL and have been using this repository to learn RLHF. I have a small question regarding the KL term on line 72 in the ppo_trainer.py file

https://github.com/tatsu-lab/alpaca_farm/blob/7ae83ae8e839a86c2c819112a242e7f2442af490/src/alpaca_farm/rl/ppo_trainer.py#L72-L73

According to equation (8) in https://arxiv.org/pdf/1707.06347.pdf, the KL penalty term is D_KL(p_ref, p). Based on the definition of KL, shouldn't line 72 be `torch.clamp(ref_logprobs - logprobs, min=0.0)`? I'm a bit confused about why it's `torch.clamp(logprobs - ref_logprobs, min=0.0)`... this seems to penalize sequences that have high likelihood if scored by ref_policy?

**Comment by lxuechen** [comment]
Hi, thanks for your question. I believe there's a misunderstanding. 

In the current RLHF implementation, the local KL term in the original PPO algorithm isn't included in the objective. 

Rather a global KL term (between the current policy model and the starting SFT model) is included to penalize the policy from deviating far away from the SFT policy. See Equation (2) in [this paper](https://arxiv.org/pdf/2203.02155.pdf). This is the (Monte Carlo estimate of the) KL being implemented in the code you quoted. 

Note the implementation maximizes the reward, which in turn (partly) minimizes the KL through the penalty term. This means `logprobs - ref_logprobs` is encouraged to be low, which then means sampled sequences are encouraged to have high `ref_logprobs`. This makes sense, as the reward model is trained on samples generated from the initial SFT model, and thus it's most suitable to rate samples close to this distribution.
